uwblueprint/focus-on-nature
# feat[BE]: add user refund flow

## Notion ticket link

[Cancellation/Refund: Create a refund request via Stripe on submit from camper select page](https://www.notion.so/uwblueprintexecs/Cancellation-Refund-Create-a-refund-request-via-Stripe-on-submit-from-camper-select-page-03d0de2661c84fb6a865b1c7cff876f6)



## Implementation description
* Created new field `RefundStatus` on the Camper obj with two states, `Paid` (default) and `Refunded`. Status changes upon going through the refund flow
* Refund is accessed through the `cancel-registration` endpoint



## Steps to test
1. Go through any registration flow and create a camper object. Navigate to MongoDB and find the generated camper id and chargeId from the payment flow. 
![image](https://github.com/uwblueprint/focus-on-nature/assets/86946736/9bdd3e03-9e9e-42a6-b16f-15f93c405d0b)

2. Using the chargeId, log into the FON Stripe account and verify that the payment was processed 
![image](https://github.com/uwblueprint/focus-on-nature/assets/86946736/7b1517fb-6412-4cf3-ab32-e39c428eb335)
![image](https://github.com/uwblueprint/focus-on-nature/assets/86946736/8ac82c44-4c43-4037-ac26-17af6bf451e9)

3. Simulate refund flow by hitting the `cancel-registration` endpoint on Postman using camper Id and chargeId as parameters 
![image](https://github.com/uwblueprint/focus-on-nature/assets/86946736/c8552636-06cd-4732-b0d7-8393919f46f1)

4. Verify camper obj `RefundStatus` has been changed to `Refunded` 
![image](https://github.com/uwblueprint/focus-on-nature/assets/86946736/ef631036-ed13-46be-9a5a-f91d954075fa)

5. Verify refund event on Stripe 
![image](https://github.com/uwblueprint/focus-on-nature/assets/86946736/6669a1ba-a52f-4c26-8d6d-05dd28c00eb3)




## What should reviewers focus on?
* What type of response to send for cancellation endpoint
* Whether partial refunds work (i.e. refunding only one camper from a chargeId with 3+ camper payment)


## Checklist
